agent/engine/docker_task_engine_test.go

```diff
@@ -3128,3 +3128,252 @@ func TestStartFirelensContainerRetryForContainerIP(t *testing.T) {
+	}
+	testTask := &apitask.Task{
+		Arn: "arn:aws:ecs:region:account-id:task/test-task-arn",
+		Containers: []*apicontainer.Container{
+			{
+				Name:      "test-container",
+				RuntimeID: "runtime-ID",
+				ExecCommandAgentMetadata: &apicontainer.ExecCommandAgentMetadata{
+					PID: execAgentPID,
+				},
+			},
+		},
+		ExecCommandAgentEnabled: true,
+	}
+	taskEngine.(*DockerTaskEngine).state.AddTask(testTask)
+	taskEngine.(*DockerTaskEngine).managedTasks[testTask.Arn] = &managedTask{Task: testTask}
+	client.EXPECT().TopContainer(gomock.Any(), testTask.Containers[0].RuntimeID, gomock.Any(), execAgentPID).Return(resp, nil)
+	taskEngine.(*DockerTaskEngine).monitorExecAgentProcesses(ctx)
+	time.Sleep(5 * time.Millisecond)
+	assert.Equal(t, execAgentPID, testTask.Containers[0].GetExecCommandAgentMetadata().PID)
+}
+
+func TestMonitorExecAgentProcessExecDisabled(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	ctrl, _, _, taskEngine, _, _, _ := mocks(t, ctx, &defaultConfig)
+	defer ctrl.Finish()
+	testTask := &apitask.Task{
+		Arn: "arn:aws:ecs:region:account-id:task/test-task-arn",
+		Containers: []*apicontainer.Container{
+			{
+				Name:      "test-container",
+				RuntimeID: "runtime-ID",
+			},
+		},
+	}
+	taskEngine.(*DockerTaskEngine).state.AddTask(testTask)
+	taskEngine.(*DockerTaskEngine).managedTasks[testTask.Arn] = &managedTask{Task: testTask}
+	taskEngine.(*DockerTaskEngine).monitorExecAgentProcesses(ctx)
+	time.Sleep(5 * time.Millisecond)
+}
+func TestMonitorExecAgentsMultipleContainers(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	ctrl, client, _, taskEngine, _, _, _ := mocks(t, ctx, &defaultConfig)
+	defer ctrl.Finish()
+	execAgentPID1 := "1234"
+	execAgentPID2 := "2345"
+	resp := &dockercontainer.ContainerTopOKBody{
+		Processes: [][]string{{"root", execAgentPID1}},
+	}
+	testTask := &apitask.Task{
+		Arn: "arn:aws:ecs:region:account-id:task/test-task-arn",
+		Containers: []*apicontainer.Container{
+			{
+				Name:      "test-container1",
+				RuntimeID: "runtime-ID1",
+				ExecCommandAgentMetadata: &apicontainer.ExecCommandAgentMetadata{
+					PID: execAgentPID1,
+				},
+			},
+			{
+				Name:      "test-container2",
+				RuntimeID: "runtime-ID2",
+				ExecCommandAgentMetadata: &apicontainer.ExecCommandAgentMetadata{
+					PID: execAgentPID2,
+				},
+			},
+		},
+		ExecCommandAgentEnabled: true,
+	}
+	taskEngine.(*DockerTaskEngine).state.AddTask(testTask)
+	taskEngine.(*DockerTaskEngine).managedTasks[testTask.Arn] = &managedTask{Task: testTask}
+	client.EXPECT().TopContainer(gomock.Any(), testTask.Containers[0].RuntimeID, gomock.Any(), execAgentPID1).Return(resp, nil)
+	client.EXPECT().TopContainer(gomock.Any(), testTask.Containers[1].RuntimeID, gomock.Any(), execAgentPID2).Return(nil, errors.New(dockerapi.TopProcessNotFoundErrorName))
+	taskEngine.(*DockerTaskEngine).monitorExecAgentProcesses(ctx)
+	time.Sleep(5 * time.Millisecond)
```

Comment:
I'm not a fan of waiting for a fixed time. This can cause flakiness when the CPU is to busy and golang runtime takes its time to schedule go routines. What do you think of creating a primitive to check for a value periodically, and have a timeout. That would make it more deterministic. 

These snippets illustrates the idea.
* This simulates a times out when the expected value is never present, which would be an error: https://play.golang.org/p/ETapQraKPMW
* This shows what happens when the condition we want actually happens: https://play.golang.org/p/rYiqsc3raVY

Reply:
+1 about the comment on waiting for a fixed time. 

This test and below one is a bit tricky, since we use go routines in the code. 
The assertion is checking if the value of the Exec Metadata's PID has been unchanged after the monitor call. 
Hence the check for a value periodically and having a timeout for that may not actually work in this case, since we will not know if the value is unchanged before/after the monitor call is over. Hence this current pattern. 
Open to better ideas.

Reply:
how about at least waiting for the `TopContainer` to be executed before the static wait? This narrows the margin for flakiness

```golang
topCtx, topCancel := context.WithTimeout(context.Background(), time.Second) 
defer topCancel()
client.EXPECT().TopContainer(gomock.Any(), testTask.Containers[1].RuntimeID, gomock.Any(),execAgentPID2).DoAndReturn(
	func(ctx context.Context, dockerID string, timeout time.Duration, psArgs ...string) (*dockercontainer.ContainerTopOKBody, error) {
		defer topCancel()
		return nil, errors.New(dockerapi.TopProcessNotFoundErrorName)
	})
taskEngine.(*DockerTaskEngine).monitorExecAgentProcesses(ctx)
<-topCtx.Done()
time.Sleep(5 * time.Millisecond)
```